o3de/o3de
# Initial Windows installer PyTests and Jenkins build config changes

~~Looking for feedback on one potential route for installer tests using PyTest.~~

Related RFC: [Proposed RFC Feature O3DE Installer Tests](https://github.com/o3de/sig-build/issues/18)


These installer tests use PyTest fixtures for dependency, to ensure some tests are run before others (e.g. wait to create a project until the installer has run, wait to run the launcher until after it is compiled)

The Jenkins build node that runs these tests needs
1. Python and PyTest
2. Git
2. Sparse Shallow clone of the folder containing the tests
3. VS Build Tools
4. Windows SDK with SignTool (optional but preferred to verify installer is signed)

Jenkins executes the tests using `pytest_windows.cmd` with a command like this
```
python -m pytest C:\git\o3de\cmake\Platform\Windows\Packaging\Tests \
    --capture=no \
    --installer-uri=s3://<bucketname>/o3de_installer.exe \
    --install-root=C:\O3DE\0.0.0.0 \
    --project-path=C:\TestProject
```

Tests run as Windows admin to avoid UAC issues.

### Tests are
1. Run installer
2. Verify key binaries exist
3. o3de.exe registers engine
4. create project succeeds
5. compile project succeeds
6. asset processor batch succeeds
7. editor runs
8. game launcher runs
9. Uninstall succeeds and unregisters engine

### Testing
- ran `installer_test` in AWS dev Jenkins instance and have had 4 consecutive green runs
- ran `installer_pipe` in AWS dev Jenkins instance successfully
- ran `installer_test` using a clean workspace and dirty workspace

### Results and issues
- incremental test run takes about 17 minutes on the target hardware c5.2xlarge, well under the initial 1 hour estimate.
- if the uninstall fails or isn't run due to an error, subsequent attempts to install O3DE may fail and you have to login to the node to manually uninstall the product or spin up a new node. This happened to me once when I cleaned the workspace but O3DE had not been uninstalled.  The installer assumed the product was already installed based on the valid registry keys, but did not know it needed to repair an existing install because the files were missing.

Signed-off-by: Alex Peterson <26804013+AMZN-alexpete@users.noreply.github.com>
